SAMZA-1100: Exception when using a stream as both bootstrap and broadcast.

shanthoosh opened the pull request:
When a task input stream is used as both broadcast and bootstrap stream in a samza job, Bootstrappingchooser marks the stream as bootstrapped when a single task finishes consuming all the SystemStreamPartitions(This happens when all the starting offset for each partition in the input stream is of type upcoming). This patch fixes this, by marking a stream as bootstrapped, only when all the systemStreamPartitions in a input stream is consumed by all the expected tasks.

More details here : https://issues.apache.org/jira/browse/SAMZA-1100

sborya commented on samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala:
```diff
@@ -354,7 +355,14 @@ object SamzaContainer extends Logging {
     val chooserFactory = Util.getObj[MessageChooserFactory](chooserFactoryClassName)
 
-    val chooser = DefaultChooser(inputStreamMetadata, chooserFactory, config, samzaContainerMetrics.registry)
+    val sspToTaskCount: mutable.Map[SystemStreamPartition, Int] = mutable.Map[SystemStreamPartition, Int]()
```
Please add a comment here explaining why we are doing it.

shanthoosh replied:
The explanation was done at multiple places, including BootstrappingChooser, DefaultChooser constructors. Hence didn't add it here. Done. Added a comment, describing the intention.

sborya commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -78,17 +85,19 @@ class BootstrappingChooser(
-          .getSystemStreamPartitionMetadata
-          .keys
-          .map(new SystemStreamPartition(systemStream, _))
+        metadata.getSystemStreamPartitionMetadata.map {
```
See if it can be rewritten more scala style :)

sborya commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -103,8 +112,8 @@ class BootstrappingChooser(
-    laggingSystemStreamPartitions = laggingSystemStreamPartitions.filter(registeredSystemStreamPartitions.contains(_))
-    systemStreamLagCounts = laggingSystemStreamPartitions.groupBy(_.getSystemStream).map {case (systemStream, ssps) => systemStream -> ssps.size}
+    laggingSystemStreamPartitions = laggingSystemStreamPartitions.filter((sspToTaskCount: (SystemStreamPartition, Int)) => registeredSystemStreamPartitions.contains(sspToTaskCount._1))
+    systemStreamLagCounts = laggingSystemStreamPartitions.keys.groupBy(_.getSystemStream).map {case (systemStream, ssps) => systemStream -> ssps.size}
```
please add comment why do we have 2 counters and what is the relationship between them.

sborya commented on samza-core/src/main/scala/org/apache/samza/system/chooser/DefaultChooser.scala:
```diff
@@ -244,7 +245,13 @@ class DefaultChooser(
+   * Stores the mapping from SystemStreamPartition to number of tasks consuming from it.
+   * This will be used by the MessageChooser to customize the envelope selection.
+   */
+  sspToTasksCount: Map[SystemStreamPartition, Int] = Map()) extends MessageChooser with Logging {
```
I think it is not a very good style to use default values in the arugments. If you don't really need them - please remove them.

sborya commented on samza-core/src/test/scala/org/apache/samza/system/chooser/TestBootstrappingChooser.scala:
```diff
@@ -193,11 +193,35 @@ class TestBootstrappingChooser(getChooser: (MessageChooser, Map[SystemStream, Sy
+    // Consume a partition from SystemStream2
+    chooser.register(envelope2.getSystemStreamPartition, "321")
+
+    assertEquals(Map(), chooser.laggingSystemStreamPartitions)
```
please also test chooser.systemStreamLagCounts.

jmakes commented on samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala:
```diff
@@ -354,7 +355,19 @@ object SamzaContainer extends Logging {
+      }
+    }
+
+    val chooser = DefaultChooser(inputStreamMetadata, chooserFactory, config, samzaContainerMetrics.registry, sspToTaskCount.toMap)
```
Something feels wrong here. Why should a chooser be concerned with which or how many tasks consume it?

I'll need to dig into the code to understand more, but I suspect that all the code for broadcast streams is on the task side, not the chooser side. If so, then we shouldn't be updating the chooser with broadcast logic.

shanthoosh replied:
Had an offline discussion. The consensus is passing in TaskCount map into the chooser will work, but it breaks the abstraction barrier and making the chooser aware of the number of downstream tasks that are consuming from it in unnecessary(violates the separation of concerns). 

The solution we'd agreed upon is retaining the lowest offset for each of the registered SystemStreamPartition and calling checkOffset for each SSP with lowest offset in BootstrappingChooser.start .

jmakes commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -120,15 +129,12 @@ class BootstrappingChooser(
     wrapped.register(systemStreamPartition, offset)
 
-    registeredSystemStreamPartitions += systemStreamPartition
+    if (!registeredSystemStreamPartitions.contains(systemStreamPartition)
```
This logic should be explained with comments.

shanthoosh replied:
Explained the reasoning in a comment and added a JIRA to comment as well. Please check if it would suffice.

jmakes commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -120,15 +129,12 @@ class BootstrappingChooser(
 
-    registeredSystemStreamPartitions += systemStreamPartition
+    if (!registeredSystemStreamPartitions.contains(systemStreamPartition)
+        || (offset != null && registeredSystemStreamPartitions(systemStreamPartition).toLong > offset.toLong)) {
```
I don't think it's safe to assume Long offsets. SystemAdmin.offsetComparator should be used.

shanthoosh replied:
Switched to use SystemAdmin.offsetComparator.

jmakes commented:
Looks like there were build errors with the latest commit. Please resolve.

shanthoosh commented:
@jmakes : 

Fixed the broken tests. 

I've added Map(SystemName -> SystemAdmins) in the classes DefaultChooser and BootstrappingChooser. In tests, those values were not set correctly. Hence the failures.

shanthoosh commented:
The failure is due to a Flaky Integration Test : TestStatefulTask. It passes in my local box. It does a kill -9 on a job in the test. Lack of permissions leads to failure.

shanthoosh commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -120,15 +133,25 @@ class BootstrappingChooser(
-    registeredSystemStreamPartitions += systemStreamPartition
+    val systemStream = systemStreamPartition.getSystem
+    val systemAdmin: SystemAdmin = systemAdmins.getOrElse(systemStream,
+                                                          throw new SamzaException("SystemAdmin is undefined for SystemStream: %s" format systemStream))
```
Fixed.

shanthoosh commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -93,7 +98,7 @@ class BootstrappingChooser(
    * Store all the systemStreamPartitions registered
    */
-  var registeredSystemStreamPartitions = Set[SystemStreamPartition]()
+  var registeredSystemStreamPartitions: mutable.Map[SystemStreamPartition, String] = mutable.Map[SystemStreamPartition, String]()
```
Removed the type annotation.

shanthoosh commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -120,15 +133,25 @@ class BootstrappingChooser(
     wrapped.register(systemStreamPartition, offset)
 
-    registeredSystemStreamPartitions += systemStreamPartition
+    val systemStream = systemStreamPartition.getSystem
```
It's `system`. Fixed.

shanthoosh commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -28,10 +27,12 @@ import org.apache.samza.metrics.MetricsHelper
-import org.apache.samza.SamzaException
 import org.apache.samza.system.SystemStreamMetadata.OffsetType
 
+import scala.collection.mutable
```
Fixed

shanthoosh commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -120,15 +133,25 @@ class BootstrappingChooser(
+     */
+    if (!registeredSystemStreamPartitions.contains(systemStreamPartition)) {
+      registeredSystemStreamPartitions += systemStreamPartition -> offset
+    } else if (offset != null) {
```
Last checkpoint for a task could be null(when it's added for the first time)  or if there is tasks addition.

shanthoosh commented:
Added a unit test that contains multiple bootstrappingChooser.register calls for same SSP with different offset(mimicking the same scenario that is the source of this issue)

nickpan47 commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -120,15 +136,24 @@ class BootstrappingChooser(
+    if (!registeredSystemStreamPartitions.contains(systemStreamPartition)) {
+      registeredSystemStreamPartitions += systemStreamPartition -> offset
+    } else if (offset != null) {
+      val comparatorResult: Integer = systemAdmin.offsetComparator(registeredSystemStreamPartitions(systemStreamPartition), offset)
```
So, what happens to the systems w/ non-comparable offsets? I would recommend to capture the UnsupportedOperationException here and wrap it w/ more meaningful logging info under this context.

shanthoosh replied:
SystemAdmin.offsetComparator returns null(as mentioned in javadoc of the interface), when the offsets are incomparable. In that case, this patch already chooses the offset provided by the first task for a ssp(lineNo: 153). 

Adding a log message at warn level. I don't want the chooser to fail & throw up the exception.

shanthoosh commented on samza-core/src/main/scala/org/apache/samza/system/chooser/BootstrappingChooser.scala:
```diff
@@ -93,7 +101,7 @@ class BootstrappingChooser(
    * Store all the systemStreamPartitions registered
    */
-  var registeredSystemStreamPartitions = Set[SystemStreamPartition]()
+  var registeredSystemStreamPartitions = mutable.Map[SystemStreamPartition, String]()
```
Fixed. Updated the javadoc description.

vjagadish1989 commented:
Merged, and submitted thanks!
